### DISCUSSION TITLE
Is it possible to define a relationship between an entity and a virtual entity?

### DISCUSSION THREAD

**Discussion by thiagomini** [mikro-orm/mikro-orm#4648] [Q&A] [answered]
Imagine the following scenario:

1. We have a table `users` with `id`, `email`, and `name`.
2. We have a table `devices` that has `user_id`, and `serial_number`
3. We want a view that joins `users.id` with `users.name` and `users.serial_number` (this is a scaled-down example just to illustrate the need of a view that uses multiple tables). Let's call that view **users_with_device**
4. We have another table `offices` that should have a relationship with `users_with_device` - that is, an office has many users with device.

I know we can model views as [virtual entities](https://mikro-orm.io/docs/virtual-entities), but is it possible to create a `1:m` relationship between an actual entity schema and a virtual entity?

**Comment by B4nan** [comment, marked as answer]
Nope, you can only have relations to actual tables, and the relations needs to be build on actual FKs, you cannot even map to another column than a primary key. Virtual entity has no primary key, no identity.

**Reply by thiagomini** [reply]
Thank you for the reply, Martin! Is there anything else we could do to accomplish that? That might be a pretty typical scenario. The definition of an `entity` for a specific context might be the union of more than a single database table, and we can do so in two ways: Using a view or a custom query (I'm considering a read-only "entity"). Could we map a column of an entity as an `expression` instead? That would already do the trick for us. I believe that might be something similar to "Formulas".

Example using the context above: 
```ts
// office.schema
users: {
  expression: (em, where) => em.createQueryBuilder()...
}
```

**Reply by B4nan** [reply]
Maybe a formula property would do it for you?

https://mikro-orm.io/docs/defining-entities#formulas

**Reply by thiagomini** [reply]
That's what I was thinking of, but can Formulas use the QueryBuilder? If so, then it should be simple enough to make it work, it would basically be a sub-select query.

**Reply by B4nan** [reply]
Oh wait a sec, and you'd like to map the formula to a (virtual) entity or some class? A formula will give you just a value, maybe it could be combined with custom type, but I'd be a bit surprised if it will just work, again not tested and never saw that in the issues either.

Maybe it could be implemented relatively easily, like a "virtual relation property", but I don't have time to experiment with this now.

**Reply by thiagomini** [reply]
Yes, I would like to map a formula (or an expression) to an entity, and use that entity as a "property" of another entity. However, to be really useful, I'd need that formula to have access to the EntityManager, just like `expressions` do. I also thought about custom types 🤔 But in any case, a "virtual relation property" could be a very useful feature! If you can point me some direction, I'd be glad to contribute to that - I mean, where do you think I should start looking at?
